src/app.py

```diff
@@ -4212,6 +4213,155 @@ def multiple_components():
+    ----------
+    entity_updates : dict
+        The dictionary of entity updates. The key is the uuid and the value is the
+        update dictionary.
+    token : str
+        The groups token for the request.
+    entity_api_url : str
+        The url of the entity-api.
+    total_tries : int, optional
+        The number of total requests to be made for each update, by default 3.
+    throttle : float, optional
+        The time to wait between requests and retries, by default 5.
+    after_each_callback : Callable[[int], None], optional
+        A callback function to be called after each update, by default None. The index
+        of the update is passed as a parameter to the callback.
+
+    Returns
+    -------
+    dict
+        The results of the bulk update. The key is the uuid of the entity. If
+        successful, the value is a dictionary with "success" as True and "data" as the
+        entity data. If failed, the value is a dictionary with "success" as False and
+        "data" as the error message.
+    """
+    headers = {
+        "Authorization": f"Bearer {token}",
+        "X-Application": "entity-api",
+    }
+    # create a session with retries
+    session = requests.Session()
+    session.headers = headers
+    retries = Retry(
+        total=total_tries,
+        backoff_factor=throttle,
+        status_forcelist=[500, 502, 503, 504],
+    )
+    session.mount(entity_api_url, HTTPAdapter(max_retries=retries))
+
+    results = {}
+    with session as s:
+        for idx, (uuid, payload) in enumerate(entity_updates.items()):
+            try:
+                # https://github.com/hubmapconsortium/entity-api/issues/698#issuecomment-2260799700
+                # yuanzhou: When you iterate over the target uuids make individual PUT /entities/<uuid> calls.
+                # The main reason we use the PUT call rather than direct neo4j query is because the entity update
+                # needs to go through the schema trigger methods and generate corresponding values programmatically
+                # before sending over to neo4j.
+                # The PUT call returns the response immediately while the backend updating may be still going on.
+                res = s.put(
+                    f"{entity_api_url}/entities/{uuid}", json=payload, timeout=15
+                )
+                results[uuid] = {
+                    "success": res.ok,
+                    "data": res.json() if res.ok else res.json().get("error"),
+                }
+            except requests.exceptions.RequestException as e:
+                logger.error(f"Failed to update entity {uuid}: {e}")
+                results[uuid] = {"success": False, "data": str(e)}
+
+            if after_each_callback:
+                after_each_callback(idx)
+
+            if idx < len(entity_updates) - 1:
+                time.sleep(throttle)
+
+    return results
+
+
+def update_datasets_uploads(entity_updates: list, token: str, entity_api_url: str) -> None:
```

Comment:
Missing function comment block above function definition.

Reply:
It's not "missing" it is in a "Docstring" below.... Please (again) see https://peps.python.org/pep-0257/, but I will move it since you have asked.